### ISSUE TITLE
Protocol violation and message receive when exceeding data limit size

### ISSUE THREAD

**Issue by tahajahangir** [aio-libs/aiosmtpd#9]
Run a simple server with data size limit:

``` python
from aiosmtpd.controller import Controller
from aiosmtpd.handlers import Message
from aiosmtpd.smtp import SMTP

class MessageHandler(Message):
    def handle_message(self, message):
        print(message)

class MyController(Controller):
    def factory(self):
        return SMTP(self.handler, data_size_limit=10)

controller = MyController(MessageHandler())
controller.start()
```

Connect to it with nc

```
$ nc -C 127.0.0.1 8025
220 hostname 1.0a1
HELO foo 
250 vpn.bayan.rocks
MAIL FROM: a@b.com
250 OK
RCPT TO: c@d.com
250 OK
DATA
354 End data with <CR><LF>.<CR><LF>
Header: value
552 Error: Too much mail data
Header2: value2
552 Error: Too much mail data
Header3: value3
552 Error: Too much mail data

552 Error: Too much mail data
.
250 OK
```

The server responds multiple time with error code 552 and finally with 250!
Also server receives message and calls `handle_message`

**Comment by warsaw** [comment]
This is still a problem although for me on git master, I get some different results.  After typing

```
Header: value
```

I see

```
Header: value
552 Error: Too much mail data
```

and then the server simply stops responding.  If I turn on debug mode (with either `PYTHONASYNCIODEBUG=1` or `-ddd` below), all I see in the console is:

```
INFO:mail.log:Data: 'DATA'
DEBUG:mail.log:b'354 End data with <CR><LF>.<CR><LF>\r\n'
DEBUG:mail.log:b'552 Error: Too much mail data\r\n'
  C-c C-cINFO:mail.log:Completed asyncio loop
```

so I don't have any idea where the server's gone off to.

Also, this is a simpler way to invoke the issue.  Put this in a file called `test.py`

```
from aiosmtpd.handlers import Message


class MessageHandler(Message):
    def handle_message(self, message):
        print(message)
```

then from a git clone:

```
$ tox -e py35-nocov --notest -r
$ PYTHONPATH=. .tox/py35-nocov/bin/smtpd -c test.MessageHandler -s 10 -ddd -l 127.0.0.1:8025 -n
```

now connect with `nc` as above.

**Comment by kozzztik** [comment]
Must be fixed with #30

**Comment by warsaw** [comment]
@kozzztik #30 is merged.  I'll test this out to see if it's still a problem, but if it is then I'll defer it post 1.0.

**Comment by pepoluan** [comment]
W.r.t. behavior ...

I can only find [this reference from z/OS](https://www.ibm.com/support/pages/smtp-receives-552-return-code-indicating-message-size-exceeds-fixed-maximum-message-size-when-sending-microsoft-mail-server). Quote:

> z/OS SMTP acts like a state machine according to the original design and RFC 821. It must finish sending the mail before accepting valid SMTP server replies. Replies received before mail has been sent will not be processed.

So... does that mean that it must reach the end first `b".\r\n" before it acts upon the 552 codes?

That begets another question: Upon receiving `b".\r\n"`, what code should the SMTP class return?

**Comment by pepoluan** [comment]
Updating this issue:

I [asked around on StackOverflow](https://stackoverflow.com/q/65088535/149900), and [a helpful individual pointed out](https://stackoverflow.com/a/65088787/149900) that [RFC 5321 section 4.2.5](https://tools.ietf.org/html/rfc5321#section-4.2.5) seems to imply that status codes are to be returned **_after_** `<CRLF>.<CRLF>`.

I think that's the closest we can get to an authoritative specification.

Hence, this issue will be closed by merging #59 (which is still in Draft stage at the moment of this posting)

**Comment by pepoluan** [comment]
#59 has exited Draft stage and is of course ready to merge. I'll wait until Monday, then if no objections, I'll merge that and (auto)close this issue.
